csrc/preseg_passes/mark_aliases_prepare.cpp

```diff
@@ -50,86 +110,100 @@ void MarkAliasesPreparePass::runPass(Fusion* fusion) {
+  // inserted.
+  //
+  // Group `uses_to_segment` by `use_of` and remove duplicates.
+  std::sort(uses_to_segment.begin(), uses_to_segment.end());
+  uses_to_segment.erase(
+      std::unique(uses_to_segment.begin(), uses_to_segment.end()),
+      uses_to_segment.end());
+
+  if (isDebugDumpEnabled(DebugDumpOption::PreSegmenterLogging)) {
+    for (const auto& use : uses_to_segment) {
+      debug() << "Will put a segment_set at " << use.user << std::endl;
     }
+  }
 
-    // If all users are used by non-alias outputs, do nothing. Adding
-    // segment_set around unlikely creates more no-op regions.
-    if (users_used_only_by_aliases.empty()) {
-      continue;
+  auto i = uses_to_segment.begin();
+  while (i != uses_to_segment.end()) {
+    TensorView* use_of = i->use_of;
+    auto j = i;
+    while (j != uses_to_segment.end() && j->use_of == use_of) {
+      j++;
     }
 
-    // If all users are used by aliases, put a `segment_set` before it.
-    if (users_used_by_non_aliases.empty()) {
-      if (aliased_io->isFusionInput()) {
-        // A `segment_set` before a fusion input is useless.
-        continue;
-      }
-      // Rarely, if `aliased_io` is already defined by `segment_set`, don't
-      // create another `segment_set`.
-      if (LoadStoreOp* def =
-              dynamic_cast<LoadStoreOp*>(aliased_io->definition())) {
-        if (def->opType() == LoadStoreOpType::SegmenterSet) {
-          continue;
+    auto insert_segment_set = [&]() {
+      // Put a `segment_set` after `use_of` and redirect aliasing users to
+      // use the `segment_set`.
+
+      // There are a few corner cases where we don't need to add a
+      // `segment_set`. If `use_of` is only used by aliases, ...
+      if (static_cast<size_t>(std::distance(i, j)) == use_of->uses().size()) {
+        if (use_of->isFusionInput()) {
+          // Putting a `segment_set` between a fusion input and its users is
+          // unnecessary.
+          return;
+        }
+
+        // Rarely, if `use_of` is already defined by `segment_set`, don't
+        // create another `segment_set`.
+        if (isSegmentSet(use_of->definition())) {
+          return;
         }
       }
-      aliased_io->cacheBefore(LoadStoreOpType::SegmenterSet);
-      continue;
-    }
 
-    // Some users of `aliased_io` are used by non-aliases, and some are used
-    // only by aliases. We put a `segment_set` after `aliased_io` and redirect
-    // the latter group to use the `segment_set`.
-    TensorView* copy = segment_set(aliased_io);
-    for (Expr* e : users_used_only_by_aliases) {
-      ir_utils::replaceValInExprInputs(e, aliased_io, copy);
-    }
-    if (aliased_io->isFusionOutput()) {
-      fusion->replaceOutput(aliased_io, copy);
-    }
+      // If all aliasing users are `segment_set`, don't create another
+      // `segment_set`.
+      if (std::all_of(
+              i, j, [](const Use& use) { return isSegmentSet(use.user); })) {
```

Comment:
this condition is `all_of`. which means, if there's only a subset of `use.user` being segmenter_set, we are still going to put another segmenter_set before them.

Reply:
Yes and that's suboptimal. I think the best fix to that is: still create another segment_set here but short circuit the double segment-set in the following std::for_each. I'll give it a shot tomorrow.

Reply:
After thinking more about this, the logic as is is probably correct because `segment_set` is supposed to block merging the edges. Consider
```
in -> (add) -> t0 -> (segment_set) -> t1 -> (reshape) -> out0
               |
               +--> (permute) -> out1
               +--> (permute) -> out2
```

IIUC, it's incorrect to redirect the `permutes` to `t1` like the following. 
```
in -> (add) -> t0 -> (segment_set) -> t1 -> (reshape) -> out0
                                      |
                                      +--> (permute) -> out1
                                      +--> (permute) -> out2
```
This is because the original `segment_set` intends to separate the `reshape` and the two `permute`s, but the new `segment_set` doesn't enforce that intention. Instead, the correct outcome should be
```
in -> (add) -> t0 -> (segment_set) -> t2 -> (segment_set) -> t1 -> (reshape) -> out0
                                      |
                                      +--> (permute) -> out1
                                      +--> (permute) -> out2
```